### ISSUE TITLE
Divide by zero in Jacobian numerical differentiation when equality bounds constraints are used

### ISSUE THREAD

**Issue by VLA05** [scipy/scipy#12502]
Hello,

It seems that in _numdiff.py module the _dense_difference function does not take into account situation of an optimisation with equal bounds constraints and where initial solution respect this constraint for some . 

In that case line 519 in _numdiff.py module dx and df are equal to zero and trigger nan in the jacobian.

It then triggers an exit n° 4 "Inequality constraints incompatible" 

#### Scipy/Numpy/Python version information:

1.5.1 1.16.5 sys.version_info(major=3, minor=7, micro=4, releaselevel='final', serial=0)

my script does not raise error with following setup

1.3.1 1.16.5 sys.version_info(major=3, minor=7, micro=4, releaselevel='final', serial=0)

**Comment by andyfaff** [comment]
This is an interesting issue.
Do you have a minimal working example? What minimizer are you using?

**Comment by andyfaff** [comment]
The code that does the numerical differentiation, `_numdiff.approx_derivative` is [designed to prevent evaluation of the target function outside bounds](https://github.com/scipy/scipy/blob/master/scipy/optimize/_numdiff.py#L13). Adjusting the step to the bounds for the situation where `lb == ub` will result in a divide by zero in `_dense_difference`.

There are a couple of things to be considered here:

- what should `approx_derivative` do if `lb == ub`? Should some code be added to detect that, and return a specific value for that particular entry? (What is that value, is it 0 or `np.inf`?) Should a warning be raised?
- In several of the minimizers there is code to [calculate jacobians](https://github.com/scipy/scipy/blob/master/scipy/optimize/slsqp.py#L280) of the constraint functions. Should those detect when constraints are at risk of sending `lb == ub` to numerical differentiation code?

**Comment by andyfaff** [comment]
I can't repeat with this example that uses scipy 1.5.1 and `'slsqp'`

```
import numpy as np
from scipy.optimize import minimize, NonlinearConstraint, rosen


def con(x):
    return x[1]

nlc = NonlinearConstraint(con, [2.], [2.])

res = minimize(rosen, [3, 2.], method='SLSQP', constraints=[nlc])
assert res.success

res = minimize(rosen, [3, 5.], method='SLSQP', constraints=[nlc])
assert res.success

def old_con(x):
    return x[1] - 2.

old_nlc = {'type': 'eq', 'fun': old_con}

res2 = minimize(rosen, [3, 5.], method='SLSQP', constraints=[old_nlc])
assert res2.success

res2 = minimize(rosen, [3, 2.], method='SLSQP', constraints=[old_nlc])
assert res2.success
```

**Comment by VLA05** [comment]
Hello,

here pseudo code used:

from scipy.optimize import minimize

def obj_fun(w, arg1, arg2, arg3):
    scalar = f(w, arg1, arg2, arg3)
return scalar

cons_slsqp = ({'type': 'eq', 'fun': con_1},
                        {'type': 'eq', 'fun': con_2, 'args':(arg1, arg2 + arg3, arg4)},
                        {'type': 'ineq', 'fun': con_3})

bounds = scipy.optimize.Bounds(np.zeros(len(w)), MAX, keep_feasible=False)

# MAX is equal to zeros for some elements and different from zero for others
# Elements of w are equal to zero at same index than MAX zeros and below MAX positive elements for others

x_rmd=minimize(obj_fun, w, jac='2-point',  args=(arg1, arg2, arg3),  method='SLSQP', constraints=cons_slsqp,                     bounds=bounds, options = {'disp': False, 'ftol': 1e-8, 'eps': 1e-5, 'maxiter': 125, 'iprint': 1} )

Error:

...\AppData\Local\Continuum\anaconda3\lib\site-packages\scipy\optimize\_numdiff.py:522: RuntimeWarning: invalid value encountered in true_divide
  J_transposed[i] = df / dx

I added following lines of code in _dense_difference  which allows the optimization to run. But a complete analysis shall be done to understand the issue:

        if (dx == 0) & (df==0) :
            J_transposed[i] = 0
        else:
            J_transposed[i] = df / dx

Regards

**Comment by VLA05** [comment]
This triggers the error

bounds = Bounds([-5, 2], [5, 2], keep_feasible=False)

res2 = minimize(rosen, [3, 5.], method='SLSQP', bounds=bounds)
assert res3.success

res2 = minimize(rosen, [3, 2.], method='SLSQP', bounds=bounds)
assert res3.success

**Comment by VLA05** [comment]
this triggers the error

bounds = Bounds([-5, 2.], [5, 2.], keep_feasible=False)

res3 = minimize(rosen, [3, 5.], method='SLSQP', bounds=bounds)
assert res3.success

res3 = minimize(rosen, [3, 2.], method='SLSQP', bounds=bounds)
assert res3.success

**Comment by andyfaff** [comment]
Things are becoming clearer now. The problem arises for elements where `bounds.lb == bounds.ub`, `approx_derivative` has a divide by zero then. In such circumstances one could rephrase the original problem by factoring out that parameter, but this will be a fairly frequent issue. I'm not sure it's the job of `approx_derivative` to deal with this. Perhaps a correction should be done inbetween the grad calculation and going back into the solver?

@mdhaber do you have any thoughts on this?

**Comment by mdhaber** [comment]
`linprog` has a presolve routine that automatically eliminates fixed variables before passing them to the solvers. So what you said - "rephrase the original problem by factoring out that parameter" - but automatic when bounds are equal (and when it is obvious from equality or inequality constraints). In `linprog`, we actually remove the associated rows of the left-hand side matrices and bake the appropriate values into the right hand side, but we can't do that for a general nonlinear problem. I suppose one way to do it here would be to wrap the user's objective and constraint functions in another function that injects the fixed parameters back into the decision variable array before the user's function is evaluated. (`scipy.stats.rv_continuous.fit` does this sort of thing, but I wouldn't follow that as an example.) There might be noticeable overhead with this, as the array would in general need to be copied to get the additional values in there, but perhaps a warning could be emitted to let the user know that they should do it manually.

**Comment by bashtage** [comment]
This feels like a bug/regression. `statsmodels` has some code that works as expected on 1.4.x but now fails on 1.5.x. At a minimum, it should be loud when lb==ub and tell users that this is not supported.

**Comment by andyfaff** [comment]
@bashtage which solver methods are you using? Do they all go through `minimize`?

The main reason strict bounds on `approx_derivative` were implemented was because there were continuous issues along the lines of *"why is my function being evaluated outside of the bounds"*. e.g. a forward derivative was being tried hard up against an upper bound. Fixing that issue causes other things to pop up.

One way to fix this is to detect when `lb==ub` and eliminate that variable, but this would require wrapper functions and might cause a slowdown. This would be a one stop change in `minimize`, but would require patching of several functions in `fmin_*`.
Warning might be a good idea, but not sure where to do that.

**Comment by sachin6061** [comment]
I got same issue in my code. to solve this problem i replace (0,0) bound to (0,1/10000000000).
this work for me.

**Comment by andyfaff** [comment]
closed by #13096
